app/workers/RulesConceptsActivity/__init__.py

```diff
@@ -304,24 +305,52 @@ def _handle_table(
         StageStatusType.IN_PROGRESS,
         scan_report_table=table,
     )
-    # handle reuse of concepts at field level
+
+    # handle reuse of concepts at field level & value level
     reuse_existing_field_concepts(table_fields, table)
-    update_job(
-        JobStageType.REUSE_CONCEPTS,
-        StageStatusType.IN_PROGRESS,
-        scan_report_table=table,
-        details="Finished at field level. Continuing at value level...",
-    )
-    # handle reuse of concepts at value level
     reuse_existing_value_concepts(table_values, table)
     update_job(
         JobStageType.REUSE_CONCEPTS,
         StageStatusType.COMPLETE,
         scan_report_table=table,
-        details="Finished",
     )
 
 
+def get_table_vocabulary_mappings(
+    data_dictionary_blob: str, table_id: int
+) -> List[Dict[str, Any]]:
+
+    # 1. Get the table name (if needed for DD lookup)
+    table = ScanReportTable.objects.get(pk=table_id)
+
+    # 2. Get field data from table_values
+    table_fields = db.get_scan_report_fields(table_id)
+
+    # 3. Extract field information, including data type
+    field_info = {
+        field["id"]: {
+            "name": field["name"],
+            "data_type": field.get("data_type"),
+        }
+        for field in table_fields
+    }
+
+    # 4. Get vocabulary mappings from the DD blob
+    _, vocab_dictionary = storage_service.get_data_dictionary(data_dictionary_blob)
+    table_vocab = vocab_dictionary.get(table.name, {})
+
+    # 5. Build the output list
+    return [
```

Comment:
I think it'd be helpful for this to be a `TypedDict` at least, similar to what's in `shared_code/models.py`. In the past the workers side of Carrot has often been dictionaries of dictionaries, and it's not easy to unpick, so it's good to add types where we can.